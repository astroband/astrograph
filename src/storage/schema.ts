// Unfortunately, DGraph schema doesn't support comments,
// so we'll remove them later before exporting, using regex
const schemaWithComments = `
  # common
  order: int @index(int) .
  key: string @index(hash) .
  kind.setOption: string .
  kind.payment: string .
  kind.pathPayment: string .
  kind.manageOffer: string .
  kind.manageDatum: string .
  kind.createAccount: string .
  kind.changeTrust: string .
  kind.allowTrust: string .

  # ledgers
  ledger.id: int @index(int) .
  ledger.prev: uid @reverse .
  version: int .
  base_fee: int .
  base_reserve: int .
  max_tx_set_size: int .
  close_time: datetime @index(hour) .
  _ingested: bool .

  # accounts
  account.id: string @index(exact) .
  deleted: bool @index(bool) .

  # assets
  asset.id: string @index(exact) .
  asset.issuer: uid @reverse .
  code: string @index(exact) .
  native: bool @index(bool) .

  # transactions
  tx.id: string @index(exact) .
  tx.ledger: uid @reverse .
  tx.source: uid @reverse .
  tx.index: int .
  tx.prev: uid @reverse .
  result_code: int .
  fee_amount: int .
  fee_charged: int .
  success: bool @index(bool) .
  time_bounds.min: dateTime .
  time_bounds.max: dateTime .
  memo.type: string .
  memo.value: string .

  # ledger change
  balance: string .

  # operations
  op.id: string @index(exact) .
  op.transaction: uid @reverse .
  op.ledger: uid @reverse .
  op.index: int .
  op.source: uid @reverse .
  op.destination: uid @reverse .
  op.kind: string @index(hash) .
  op.prev: uid @reverse .
  amount: int .

  # account merge
  result: uid .
  account_merge_op.result_code: int .

  # allow trust
  allow_trust_op.trustor: uid @reverse .
  allow_trust_op.result_code: int .
  allow_trust_op.asset: uid @reverse .
  authorize: bool @index(bool) .

  # bump sequence
  bump_sequence_op.result_code: int .
  bump_to: int @index(int) .

  # change trust
  change_trust_op.asset: uid @reverse .
  change_trust_op.result_code: int .
  limit: int @index(int) .

  # create account
  starting_balance: int @index(int) .
  create_account_op.result_code: int .

  # manage data
  manage_data_op.result_code: int .
  name: string @index(exact) .
  value: string @index(exact) . 

  # manage offer
  manage_offer_op.asset_buying: uid @reverse .
  manage_offer_op.asset_selling: uid @reverse .
  manage_offer_op.result_code: int .
  price_n: int .
  price_d: int .
  price: float .
  offer_id: int @index(int) .

  # payment
  payment_op.asset: uid @reverse .
  payment_op.result_code: int .

  # path payment
  path_payment_op.asset_source: uid @reverse .
  path_payment_op.asset_destination: uid @reverse .
  path_payment_op.assets_path: uid @reverse .
  path_payment_op.result_code: int .
  send_max: int .
  
  # set options
  set_options_op.signer: uid @reverse .
  set_options_op.inflation_destination: uid @reverse .
  set_options_op.result_code: int .
  set_options_op.signer: uid .
  clear_flags: int .
  set_flags: int .
  master_weight: int .
  home_domain: string .
  thresholds: uid .

  # thresholds
  high: int .
  med: int .
  low: int .

  # signer
  account: uid .
  weight: int .

  # operation results
  # account merge result
  source_account_balance: int .

  # path payment result
  path_payment_result.no_issuer: uid .
  path_payment_result.last: uid .
  destination: uid .
  last.asset: uid .
  path_payment_result.offers: uid .
  asset_sold: uid .
  asset_bought: uid .
  amount_sold: int .
  amount_bought: int .
  seller: uid .
`;

// remove comments
const validDgraphSchema = schemaWithComments.replace(/#.+$/gm, "");

export { validDgraphSchema as SCHEMA };
