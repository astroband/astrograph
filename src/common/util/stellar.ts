import Big from "big.js";
import { Keypair, Network } from "stellar-base";
import db from "../../database";
import { Asset } from "../../model";

const StellarAmountPrecision = 7;

export const NATIVE_ASSET_CODE = "XLM";

export async function setNetwork(): Promise<string> {
  const promise = db.storeState.getStellarNetworkPassphrase().then((networkPassphrase: string) => {
    Network.use(new Network(networkPassphrase));
    return networkPassphrase;
  });

  return promise;
}

export function buildNativeAsset(): Asset {
  return new Asset(true, NATIVE_ASSET_CODE, Keypair.master().publicKey());
}

// converts amounts according to Stellar precision like this:
// "99999999800" -> "9999.9999800"
export function toFloatAmountString(intAmountString: string): string {
  const floatAmount = new Big(intAmountString);
  return floatAmount.div(new Big("1e" + StellarAmountPrecision)).toFixed(StellarAmountPrecision);
}
