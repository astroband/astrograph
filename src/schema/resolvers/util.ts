import { Asset, Memo, MemoText, MemoHash, MemoReturn } from "stellar-sdk";
import { createBatchResolver as create } from "graphql-resolve-batch";
import { Ledger, MutationType } from "../../model";

export function createBatchResolver<T, R>(loadFn: any) {
  return create<T, R>(async (source: ReadonlyArray<T>, args: any, context: any) => loadFn(source, args, context));
}

export function ledgerResolver(obj: any) {
  const seq = obj.lastModified || obj.ledgerSeq;
  return new Ledger(seq);
}

export function memoResolver(obj: any) {
  const memo = obj.memo as Memo;
  let value = memo.value;

  if (value instanceof Buffer) {
    switch(memo.type) {
      case MemoText:
        value = value.toString("utf8");
        break;
      case MemoHash:
      case MemoReturn:
        value = value.toString("base64");
        break;
    }
  }

  return {
    type: memo.type,
    value: value
  }
}

export function assetResolver(obj: any) {
  const asset = obj.asset as Asset;
  return {
    code: asset.getCode(),
    issuer: asset.getIssuer(),
    native: asset.isNative()
  }
}

export function eventMatches(args: any, id: string, mutationType: MutationType): boolean {
  const idEq: boolean | null = args.idEq ? id === args.idEq : null;
  const idIn: boolean | null = args.idIn ? args.idIn.includes(id) : null;
  const mutationTypeIn: boolean | null = args.mutationTypeIn ? args.mutationTypeIn.includes(mutationType) : null;

  const conditions = [idEq, idIn, mutationTypeIn].filter(c => c !== null);

  return conditions.every(c => c === true);
}
