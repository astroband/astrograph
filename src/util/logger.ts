import { createLogger, format, Logger, transports } from "winston";

enum LoggerLevel {
  error = "error",
  warn = "warn",
  info = "info",
  verbose = "verbose",
  debug = "debug",
  silly = "silly"
}
const PROD_LOG_LEVEL = LoggerLevel.debug;
const DEV_LOG_LEVEL = LoggerLevel.debug;
const LOG_FILE = LoggerLevel.error;

let logLevel = process.env.LOG_LEVEL;

if (!logLevel || !LoggerLevel[logLevel]) {
  logLevel = process.env.NODE_ENV === "production" ? PROD_LOG_LEVEL : DEV_LOG_LEVEL;
}
const errorWithStacktraceFormat = format(info => {
  if (info instanceof Error) {
    return { message: info.message, stack: info.stack, ...info };
  }

  return info;
});

const logger: Logger = createLogger({
  format: format.combine(
    errorWithStacktraceFormat(),
    format.colorize(),
    format.splat(),
    format.timestamp(),
    format.simple()
    // format.printf(info => `${info.timestamp} ${info.level}: ${info.message} Stacktrace: ${info.stack}`)
  ),
  transports: [
    new transports.Console({ level: logLevel }),
    new transports.File({ filename: "error.log", level: LOG_FILE })
  ]
});

if (process.env.NODE_ENV !== "production") {
  logger.debug("Logging initialized at debug level");
}
export default logger;
