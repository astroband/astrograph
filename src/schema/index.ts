import { makeExecutableSchema, mergeSchemas } from "apollo-server";

import resolvers from "./resolvers";
import typeDefs from "./type_defs";

const schema = makeExecutableSchema({
  typeDefs,
  resolverValidationOptions: {
    requireResolversForResolveType: false
  }
});
const schemas = [schema];

export default mergeSchemas({ schemas, resolvers });
