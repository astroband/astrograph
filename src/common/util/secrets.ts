import dotenv from "dotenv";
import fs from "fs";
import logger from "./logger";

const envFile = process.env.NODE_ENV ? `.env.${process.env.NODE_ENV}` : ".env";

if (fs.existsSync(envFile)) {
  logger.log("info", `Using ${envFile} file to supply config environment variables`);
  dotenv.config({ path: envFile });
}

export const DB = process.env.DB || "stellar";
export const DBPORT = Number.parseInt(process.env.DBPORT || "5432", 10);
export const DBHOST = process.env.DBHOST;
export const DBUSER = process.env.DBUSER || "stellar";
export const DBPASSWORD = process.env.DBPASSWORD || "";

export const PORT = Number.parseInt(process.env.PORT || "4000", 10);
export const BIND_ADDRESS = process.env.BIND_ADDRESS || "0.0.0.0";

export const DEBUG_LEDGER = Number.parseInt(process.env.DEBUG_LEDGER || "", 10);
export const INGEST_INTERVAL = Number.parseInt(process.env.INGEST_INTERVAL || "", 10) || 2000;
