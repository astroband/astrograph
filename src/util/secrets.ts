import dotenv from "dotenv";
import fs from "fs";
import logger from "./logger";

const environment = process.env.NODE_ENV;
const envFile = process.env.NODE_ENV ? `.env.${process.env.NODE_ENV}` : ".env";

if (fs.existsSync(envFile)) {
  logger.log("info", `Using ${envFile} file to supply config environment variables`);
  dotenv.config({ path: envFile });
} else if (environment === "test" && !process.env.CI) {
  throw new Error("No .env file found for the test environment. Create `.env.test` file with necessary settings");
}

export const DB = process.env.DB || "stellar";
export const DBPORT = Number.parseInt(process.env.DBPORT || "5432", 10);
export const DBHOST = process.env.DBHOST;
export const DBUSER = process.env.DBUSER || "stellar";
export const DBPASSWORD = process.env.DBPASSWORD || "";

export const PORT = Number.parseInt(process.env.PORT || "4000", 10);
export const BIND_ADDRESS = process.env.BIND_ADDRESS || "0.0.0.0";

export const DEBUG_LEDGER = Number.parseInt(process.env.DEBUG_LEDGER || "", 10);
export const INGEST_INTERVAL = Number.parseInt(process.env.INGEST_INTERVAL || "", 10) || 2000;

export const DGRAPH_QUERY_URL = process.env.DGRAPH_QUERY_URL || process.env.DGRAPH_URL;
export const DGRAPH_INGEST_URL = process.env.DGRAPH_INGEST_URL || process.env.DGRAPH_URL;

export const STELLAR_NETWORK = process.env.STELLAR_NETWORK || "pubnet";
export const SENTRY_DSN = process.env.SENTRY_DSN;
