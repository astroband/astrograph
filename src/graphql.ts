import "./util/memo";

import * as Sentry from "@sentry/node";

import { ApolloError, ApolloServer } from "apollo-server";
import { GraphQLError } from "graphql";

import { Cursor, Worker } from "./ingest";
import schema from "./schema";
import { Connection } from "./storage/connection";
import logger from "./util/logger";
import { BIND_ADDRESS, DEBUG_LEDGER, DGRAPH_QUERY_URL, INGEST_INTERVAL, PORT, SENTRY_DSN } from "./util/secrets";
import { setNetwork as setStellarNetwork } from "./util/stellar";

if (SENTRY_DSN) {
  Sentry.init({
    dsn: SENTRY_DSN,
    integrations: [new Sentry.Integrations.RewriteFrames({ root: __dirname || process.cwd() })]
  });
}

const server = new ApolloServer({
  schema,
  tracing: true,
  introspection: true,
  playground: true,
  debug: true,
  cors: true,
  formatError: (error: ApolloError) => {
    logger.error(error);

    if (!error.originalError || error.originalError.constructor.name !== "UserInputError") {
      Sentry.captureException(error);
    }

    return new GraphQLError(
      error.message,
      error.nodes,
      error.source,
      error.positions,
      error.path,
      error,
      error.extensions
    );
  }
});

const network = setStellarNetwork();
logger.info(`Using ${network}`);

if (DGRAPH_QUERY_URL) {
  logger.info(`[DGraph] Updating schema...`);
  new Connection().migrate().catch((err: any) => {
    logger.error(err);
    Sentry.captureException(err);
    process.exit(-1);
  });
}

Cursor.build(DEBUG_LEDGER).then(cursor => {
  logger.info(
    `Staring ingest every ${INGEST_INTERVAL} ms. from ${
      DEBUG_LEDGER === -1 ? "first ledger" : DEBUG_LEDGER || "lastest ledger"
    }`
  );

  const tick = async () => {
    const worker = new Worker(cursor);
    worker
      .run()
      .then(done => {
        logger.info(`Ledger ${cursor.current}: done.`);
        if (done) {
          tick();
        } else {
          setTimeout(tick, INGEST_INTERVAL);
        }
      })
      .catch(e => {
        logger.error(e);
        if (e.message.includes("Please retry again, server is not ready to accept requests")) {
          setTimeout(tick, 200);
          return;
        }

        Sentry.captureException(e);
      });
  };

  tick();
});

server.listen({ port: PORT, host: BIND_ADDRESS }).then(({ url }) => {
  logger.info(`ğŸš€ Server ready at ${url}`);
});
