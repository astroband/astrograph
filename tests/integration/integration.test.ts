import { execSync } from "child_process";
import fs from "fs";
import { graphql } from "graphql";
import path from "path";
import { Client as dbClient } from "pg";
import { Network } from "stellar-base";
import schema from "../../src/schema";
import logger from "../../src/util/logger";
import * as secrets from "../../src/util/secrets";

Network.useTestNetwork();

const testCases = [
  "Assets",
  "Single account query",
  "Accounts signed by",
  "Ledgers",
  "Trust lines"
];

async function importDbDump() {
  const client = new dbClient({
    host: secrets.DBHOST,
    port: secrets.DBPORT,
    database: secrets.DB,
    user: secrets.DBUSER,
    password: secrets.DBPASSWORD
  });

  const sql = fs.readFileSync(path.join(__dirname, "test_db.sql"), "utf8");

  await client.connect();

  logger.log("info", "importing database fixture...");

  await client.query(sql);
  await client.end();
}

describe("Integration tests", () => {
  beforeAll(async () => {
    try {
      await importDbDump();
    } catch (e) {
      if (e.message !== `database "${secrets.DB}" does not exist`) {
        throw e;
      }

      logger.log("info", `${e.message}. Creating...`);
      execSync(`createdb ${secrets.DB}`);
      await importDbDump();
    }
  });

  test.each(testCases)("%s", async (caseName: string) => {
    const queryFile = caseName.toLowerCase().replace(/ /g, "_");
    const query = fs.readFileSync(`${__dirname}/integration_queries/${queryFile}.gql`, "utf8");

    const { data } = await graphql(schema, query);

    expect(data).toMatchSnapshot();
  });
});
