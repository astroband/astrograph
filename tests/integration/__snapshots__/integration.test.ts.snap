// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests Accounts signed by 1`] = `
Object {
  "accountsSignedBy": Array [
    Object {
      "id": "GCVIRZIN4CGYY56CSL7RYAFHKQTGE34GIASZ2D4IGZGV3FDL622LPQZT",
      "ledger": Object {
        "seq": 4,
      },
      "signers": Array [
        Object {
          "account": Object {
            "id": "GCVIRZIN4CGYY56CSL7RYAFHKQTGE34GIASZ2D4IGZGV3FDL622LPQZT",
          },
          "weight": 1,
        },
      ],
      "trustLines": Array [
        Object {
          "asset": Object {
            "code": "XLM",
          },
          "authorized": true,
        },
      ],
    },
    Object {
      "id": "GDQWLXN6B7IVBJ2Z2DB2VAEHYEJT4POZBB5DVPL2GWJ67YRSYPLTZ6WC",
      "ledger": Object {
        "seq": 3,
      },
      "signers": Array [
        Object {
          "account": Object {
            "id": "GDQWLXN6B7IVBJ2Z2DB2VAEHYEJT4POZBB5DVPL2GWJ67YRSYPLTZ6WC",
          },
          "weight": 1,
        },
        Object {
          "account": Object {
            "id": "GDQWLXN6B7IVBJ2Z2DB2VAEHYEJT4POZBB5DVPL2GWJ67YRSYPLTZ6WC",
          },
          "weight": 1,
        },
      ],
      "trustLines": Array [
        Object {
          "asset": Object {
            "code": "XLM",
          },
          "authorized": true,
        },
      ],
    },
  ],
}
`;

exports[`Integration tests Data entries 1`] = `
Object {
  "dataEntries": Array [
    Object {
      "account": Object {
        "id": "GCVIRZIN4CGYY56CSL7RYAFHKQTGE34GIASZ2D4IGZGV3FDL622LPQZT",
      },
      "ledger": Object {
        "header": Object {
          "ledgerVersion": 10,
        },
        "seq": 4,
      },
      "value": "test_value",
    },
  ],
}
`;

exports[`Integration tests Ledgers 1`] = `
Object {
  "ledgers": Array [
    Object {
      "header": Object {
        "ledgerVersion": 10,
      },
      "seq": 2,
    },
    Object {
      "header": Object {
        "ledgerVersion": 10,
      },
      "seq": 3,
    },
    Object {
      "header": Object {
        "ledgerVersion": 10,
      },
      "seq": 4,
    },
  ],
}
`;

exports[`Integration tests Signers 1`] = `
Object {
  "signers": Array [
    Object {
      "account": Object {
        "id": "GDQWLXN6B7IVBJ2Z2DB2VAEHYEJT4POZBB5DVPL2GWJ67YRSYPLTZ6WC",
      },
      "signer": Object {
        "id": "GDQWLXN6B7IVBJ2Z2DB2VAEHYEJT4POZBB5DVPL2GWJ67YRSYPLTZ6WC",
      },
      "weight": 1,
    },
    Object {
      "account": Object {
        "id": "GDQWLXN6B7IVBJ2Z2DB2VAEHYEJT4POZBB5DVPL2GWJ67YRSYPLTZ6WC",
      },
      "signer": Object {
        "id": "GCVIRZIN4CGYY56CSL7RYAFHKQTGE34GIASZ2D4IGZGV3FDL622LPQZT",
      },
      "weight": 1,
    },
  ],
}
`;

exports[`Integration tests Single account query 1`] = `
Object {
  "account": Object {
    "data": Array [
      Object {
        "name": "test_data",
        "value": "test_value",
      },
    ],
    "flags": Object {
      "authImmutable": false,
      "authRequired": false,
      "authRevokable": false,
    },
    "id": "GCVIRZIN4CGYY56CSL7RYAFHKQTGE34GIASZ2D4IGZGV3FDL622LPQZT",
    "inflationDest": null,
    "ledger": Object {
      "header": Object {
        "ledgerVersion": 10,
      },
      "seq": 4,
    },
    "sequenceNumber": "8589934593",
    "signerFor": Array [
      Object {
        "id": "GCVIRZIN4CGYY56CSL7RYAFHKQTGE34GIASZ2D4IGZGV3FDL622LPQZT",
        "trustLines": Array [
          Object {
            "asset": Object {
              "code": "XLM",
              "issuer": null,
              "native": true,
            },
          },
        ],
      },
      Object {
        "id": "GDQWLXN6B7IVBJ2Z2DB2VAEHYEJT4POZBB5DVPL2GWJ67YRSYPLTZ6WC",
        "trustLines": Array [
          Object {
            "asset": Object {
              "code": "XLM",
              "issuer": null,
              "native": true,
            },
          },
        ],
      },
    ],
    "signers": Array [
      Object {
        "signer": Object {
          "id": "GCVIRZIN4CGYY56CSL7RYAFHKQTGE34GIASZ2D4IGZGV3FDL622LPQZT",
        },
      },
    ],
    "thresholds": Object {
      "high": 0,
      "low": 0,
      "masterWeight": 1,
      "medium": 0,
    },
    "trustLines": Array [
      Object {
        "account": Object {
          "id": "GCVIRZIN4CGYY56CSL7RYAFHKQTGE34GIASZ2D4IGZGV3FDL622LPQZT",
        },
        "asset": Object {
          "code": "XLM",
          "issuer": null,
          "native": true,
        },
      },
    ],
  },
}
`;

exports[`Integration tests Trust lines 1`] = `
Object {
  "trustLines": Array [
    Object {
      "asset": Object {
        "code": "XLM",
        "native": true,
      },
      "authorized": true,
      "balance": "999.9999900",
    },
    Object {
      "asset": Object {
        "code": "KHL",
        "native": false,
      },
      "authorized": true,
      "balance": "0.0000000",
    },
  ],
}
`;
